File: go/tree/leetcode/kth_smallest_element_bst.go
package leetcode

// 二叉搜索树中第k小的元素

// 中序遍历取第k个
func kthSmallest1(root *TreeNode, k int) int {
	var inorder func(node *TreeNode)
	var vals []int
	inorder = func(node *TreeNode) {
		if node == nil {
			return
		}
		inorder(node.Left)
		vals = append(vals, node.Val)
		inorder(node.Right)
	}

	inorder(root)
	return vals[k-1]
}

// 中序遍历优化，遍历到第k个时就是结果
func kthSmallest2(root *TreeNode, k int) int {
	var stack []*TreeNode
	var cur = root
	for len(stack) > 0 || cur != nil {
		for cur != nil {
			stack = append(stack, cur)
			cur = cur.Left
		}
		node := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		k--
		if k == 0 {
			return node.Val
		}
		cur = node.Right
	}
	return 0
}

File: cpp/tree/kth_smallest_element_bst.cpp
// 二叉搜索树中第k小的元素

#include "stack"
#include "queue"
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() {}
    TreeNode(int val) : val(val) {}
    TreeNode(int val, TreeNode *left, TreeNode *right) : val(val), left(left), right(right) {}
};

class Solution {
public:

    // 中序遍历，取第k个
    int kthSmallest1(TreeNode* root, int k) {
        vector<int> vals;
        helper(root, vals);
        return vals[k - 1];
    }

    void helper(TreeNode* root, vector<int>& vals) {
        if (!root) {
            return;
        }
        helper(root->left, vals);
        vals.push_back(root->val);
        helper(root->right, vals);
    }

    // 中序遍历优化，取到第k个时结束
    int kthSmallest2(TreeNode* root, int k) {
        stack<TreeNode*> s;
        TreeNode* cur = root;
        while (!s.empty() || cur) {
            while (cur) {
                s.push(cur);
                cur = cur->left;
            }
            TreeNode* node = s.top();
            s.pop();
            if (--k == 0) {
                return node->val;
            }
            cur = node->right;
        }
        return 0;
    }
};
